subsystem/src/lib.rs

```diff
@@ -0,0 +1,497 @@
+    }
+
+    #[allow(unused)]
+    async fn wait_for_shutdown_with_timeout(
+        name: &'static str,
+        shutting_down_rx: mpsc::Receiver<()>,
+        timeout: Option<Duration>,
+    ) {
+        let shutdown_future = Self::wait_for_shutdown(shutting_down_rx);
+        if let Some(timeout) = timeout {
+            cfg_if::cfg_if! {
+                if #[cfg(all(feature = "time", not(loom)))] {
+                    // Wait for shutdown under a timeout
+                    if let Err(elapsed) = tokio::time::timeout(timeout, shutdown_future).await {
+                        log::error!("Manager {} shutdown timed out", name);
+                    }
+                } else {
+                    // Timeout was requested but is not supported
+                    if cfg!(not(feature = "time")) {
+                        log::error!("Shutdown timeout support not compiled in");
+                    } else if cfg!(loom) {
+                        log::warn!("Shutdown timeout disabled under loom");
+                    }
+                    shutdown_future.await
+                }
+            }
+        } else {
+            // No timeout requested, just wait for shutdown
+            shutdown_future.await
+        }
+    }
+
+    /// Run the application main task.
+    ///
+    /// Completes when all the subsystems are fully shut down.
+    pub async fn main(mut self) {
+        // The main task just waits for the shutdown signal and coordinates the cleanup at the end.
+        // All other functionality is performed by subsystems.
+
+        log::info!("Manager {} running", self.name);
+
+        // Signal the manager is shut down so it does not wait for itself
+        std::mem::drop(self.shutting_down_tx);
+
+        // Wait for a subsystem to shut down.
+        self.shutting_down_rx
+            .recv()
+            .await
+            .unwrap_or_else(|| log::info!("Manager {}: all subsystems already down", self.name));
+
+        log::info!("Manager {} shutting down", self.name);
+
+        // Order all the remaining subsystems to shut down.
+        let _ = self.shutdown_request_tx.send(());
+
+        // Wait for the subsystems to go down.
+        Self::wait_for_shutdown_with_timeout(
+            self.name,
+            self.shutting_down_rx,
+            self.shutdown_timeout,
+        )
+        .await;
+
+        log::info!("Manager {} terminated", self.name);
+    }
+}
+
+/// Subsystem configuration
+pub struct SubsystemConfig {
+    /// Subsystem name
+    pub subsystem_name: &'static str,
+    /// Capacity of the call request channel
+    pub call_queue_capacity: usize,
+}
+
+impl SubsystemConfig {
+    const DEFAULT_CALL_QUEUE_CAPACITY: usize = 64;
+    const DEFAULT_SUBSYSTEM_NAME: &'static str = "<unnamed>";
+
+    /// New configuration with given name, all other options are defaults.
+    fn named(subsystem_name: &'static str) -> Self {
+        Self {
+            subsystem_name,
+            ..Default::default()
```

Comment:
I get that the intention of this function is to just define a configuration just by name. However, I'm allergic to these blanket defaults... I always imagine someone adding some other configuration field, and then it getting assigned by default and causing problems. I would rather get a compiler error, and have the developer attend to it and choose the correct value instead of a possible unintended default that could create problems.

My ask: Please change this to: `call_queue_capacity: Default::default()`.

Reply:
Then please set a default value in a different way.

Reply:
How about now?